feat(quickfix): handle name display in qflist for hard links

austin880625 opened the pull request:
It is a small enhancement on quickfix list. It currently shows entries with duplicate name if the file is opened with the path of hard links.

This behavior can be produced by creating a file like `main1.cpp` and make its hard link `main2.cpp` then issue the following commands in vim:

```
:call setqflist([{'filename':'main1.cpp', 'lnum':1}, {'filename':'main2.cpp', 'lnum':1}])
:copen
``` 

The major cause is that qflist assumed that the filename passed into `qf_add_entry` matches the filename opened with the buffer.

This patch handles this case by introduce a `qf_fname` into `qfline_S` structure. It stores the filename from `qf_add_entry` for each quickfix line.

I'm not really familiar with vim codebase and I'm redirected from [nvim](https://github.com/neovim/neovim/pull/30368). Please point out if this PR would introduce any issues, thanks!

yegappan commented on src/quickfix.c:
```diff
@@ -36,6 +36,7 @@ struct qfline_S
     int		qf_end_col;	// column when the error has range or zero
     int		qf_nr;		// error number
     char_u	*qf_module;	// module name for this error
+    char_u	*qf_fname;	// filename for a line
```
Creating a copy of the filename for each quickfix entry will significantly increase the memory used for the quickfix list.  Is this really needed?

austin880625 replied:
Thanks for the comment. My bad that I wasn't aware of the memory usage at the time. My thought was that the list is from LSP response. If the filenames already occupy the memory before `setqflist`, then it should be ok to stay there(copy & release the original one by caller) and hold by qflist.

Another approach would be doing a comparison with `buf->b_ffname` and only copy the filename if they are different, or add an option on `setqflist` to tell if caller want it to do so. This still leads to another doubt that if the time of comparison on each entry really worth.

yegappan commented on src/quickfix.c:
```diff
@@ -2175,6 +2176,8 @@ qf_add_entry(
 {
     qfline_T	*qfp;
     qfline_T	**lastp;	// pointer to qf_last or NULL
+    char_u	*fullname = fix_fname(fname);
```
Both filenames refer to the same file.  So I am not sure whether this is a significant issue that should be addressed.

austin880625 replied:
I believe my use case(LSP + a build infrastructure that makes hard links) is not very common. I just saw it and try to see if I can make it better. This issue makes me seeing alot of duplicated entries when looking on references of a function. If they should be treated as same file, then the entries shouldn't be duplicated. But this also depends on which component we think should handle this case:

1. vim: This is what my patch tries to deals with.
2. the plugin or LSP: LSP shouldn't send two names in the responses if they are already the same files.
3. Users:  Hard links are OS specific and just make a lot of mess in general. Users shouldn't expect editors/LSPs/plugins to know well on how files are linked.

As I see that vim only shows one filename when the file's hard link is opened(instead of showing another name). I believe this implementation was thought(and tested) more extensively and the better ways to address this issue might be 2. or 3. If that's the case then we can just close this PR.

austin880625 committed:
feat(quickfix): handle name display in qflist for hard links

Quickfix list shows entries with duplicate name if the file is opened
with the path of hard links.

The major cause is that qflist assumed that the filename passed into
`qf_add_entry` matches the filename opened with the buffer.

This patch handles this case by introduce a `qf_fname` into `qfline_S`
structure. It stores the filename from `qf_add_entry` for each quickfix
line.

yegappan commented on src/testdir/test_quickfix.vim:
```diff
@@ -6578,4 +6578,32 @@ func Test_cbuffer_range()
 
+" Test for displaying fname pass from setqflist when the name
+" are hard links to prevent seemly duplicate entries.
+func Xtest_hardlink_fname(cchar)
```
You need to add another function starting with the prefix "Test_" that calls this function with the argument 'c' and 'l'.

yegappan commented on src/testdir/test_quickfix.vim:
```diff
@@ -6578,4 +6578,32 @@ func Test_cbuffer_range()
+    void sample() {}
+    int main() { sample(); return 0; }
+  END
+  call writefile(lines, 'file1.c')
```
Pass the flag 'D' as third argument  to this function.

yegappan replied:
Use a more unique file name than "file1.c" (e.g. Xtesthardlink.c")

yegappan commented on src/testdir/test_quickfix.vim:
```diff
@@ -6578,4 +6578,32 @@ func Test_cbuffer_range()
+  " like how LSP plugins set response into qflist
+  call g:Xsetlist([{'filename' : 'file1.c', 'lnum' : 1},
+        \ {'filename' : 'file2.c', 'lnum' : 1}], ' ')
+  copen
```
Use "Xopen" instead of "copen".

yegappan commented on src/testdir/test_quickfix.vim:
```diff
@@ -6578,4 +6578,32 @@ func Test_cbuffer_range()
+  " so that they aren't seen as duplication.
+  call assert_equal(['file1.c|1| ',
+        \ 'file2.c|1| '], getline(1, '$'))
+  cclose
```
Use "Xclose" instead of "cclose".

resslr reviewed (commented):
accidentally reviewed @yegappan delete this comment please

zeertzjq commented on src/testdir/test_quickfix.vim:
```diff
@@ -6578,4 +6578,37 @@ func Test_cbuffer_range()
+" Test for displaying fname pass from setqflist when the name
+" are hard links to prevent seemly duplicate entries.
+func Xtest_hardlink_fname(cchar)
+  CheckUnix
```
It may be better to move this line to the start of `Test_hardlink_fname()`

austin880625 replied:
updated, thanks
